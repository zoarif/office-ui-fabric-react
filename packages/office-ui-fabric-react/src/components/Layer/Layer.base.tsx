/* tslint:disable:no-unused-variable */
import * as React from 'react';
import * as ReactDOM from 'react-dom';
/* tslint:enable:no-unused-variable */

import { Fabric } from '../../Fabric';
import {
  ILayerProps,
  ILayerStyleProps,
  ILayerStyles,
} from './Layer.types';
import {
  BaseComponent,
  classNamesFunction,
  customizable,
  getDocument,
  setVirtualParent,
  createRef
} from '../../Utilities';

const _layersByHostId: { [hostId: string]: LayerBase[] } = {};
let _defaultHostSelector: string | undefined;

const getClassNames = classNamesFunction<ILayerStyleProps, ILayerStyles>();

@customizable('Layer', ['theme', 'hostId'])
export class LayerBase extends BaseComponent<ILayerProps, {}> {

  public static defaultProps: ILayerProps = {
    onLayerDidMount: () => undefined,
    onLayerWillUnmount: () => undefined
  };

  private _rootElement = createRef<HTMLDivElement>();
  private _host: Node;
  private _layerElement: HTMLElement | undefined;
  private _hasMounted: boolean;
  /**
   * Used for notifying applicable Layers that a host is available/unavailable and to re-evaluate Layers that
   * care about the specific host.
   */
  public static notifyHostChanged(id: string) {
    if (_layersByHostId[id]) {
      _layersByHostId[id].forEach(layer => layer.forceUpdate());
    }
  }

  /**
   * Sets the default target selector to use when determining the host in which
   * Layered content will be injected into. If not provided, an element will be
   * created at the end of the document body.
   *
   * Passing in a falsey value will clear the default target and reset back to
   * using a created element at the end of document body.
   */
  public static setDefaultTarget(selector?: string) {
    _defaultHostSelector = selector;
  }

  constructor(props: ILayerProps) {
    super(props);

    this._warnDeprecations({
      onLayerMounted: 'onLayerDidMount'
    });

    if (this.props.hostId) {
      if (!_layersByHostId[this.props.hostId]) {
        _layersByHostId[this.props.hostId] = [];
      }

      _layersByHostId[this.props.hostId].push(this);
    }
  }

  public componentDidMount(): void {
    this.componentDidUpdate();
  }

  public componentWillUnmount(): void {
    this._removeLayerElement();

    if (this.props.hostId) {
      _layersByHostId[this.props.hostId] = _layersByHostId[this.props.hostId].filter(layer => layer !== this);
      if (!_layersByHostId[this.props.hostId].length) {
        delete _layersByHostId[this.props.hostId];
      }
    }
  }

  public componentDidUpdate(): void {
    const host = this._getHost();

    const { className, getStyles, theme } = this.props;
    const classNames = getClassNames(getStyles!,
      {
        theme: theme!,
        className,
        isNotHost: !this.props.hostId
      }
    );

    if (host !== this._host) {
      this._removeLayerElement();
    }

    if (host) {
      this._host = host;

      if (!this._layerElement) {
        const rootElement = this._rootElement.current;
        const doc = getDocument(rootElement);

        if (!doc || !rootElement) {
          return;
        }

        this._layerElement = doc.createElement('div');
        this._layerElement.className = classNames.root!;

        host.appendChild(this._layerElement);
        setVirtualParent(this._layerElement, rootElement);
      }

      // Using this 'unstable' method allows us to retain the React context across the layer projection.
      ReactDOM.unstable_renderSubtreeIntoContainer(
        this,
        (
          <Fabric className={ classNames.content }>
            { this.props.children }
          </Fabric>
        ),
        this._layerElement,
        () => {
          if (!this._hasMounted) {
            this._hasMounted = true;

            // TODO: @deprecated cleanup required.
            if (this.props.onLayerMounted) {
              this.props.onLayerMounted();
            }

            this.props.onLayerDidMount!();
          }
        });
    }
  }

  public render(): JSX.Element {
    return (
      <span
        className='ms-Layer'
        ref={ this._rootElement }
      />
    );
  }

  private _removeLayerElement(): void {
    if (this._layerElement) {
      this.props.onLayerWillUnmount!();

      ReactDOM.unmountComponentAtNode(this._layerElement);
      const parentNode = this._layerElement.parentNode;
      if (parentNode) {
        parentNode.removeChild(this._layerElement);
      }
      this._layerElement = undefined;
      this._hasMounted = false;
    }
  }

  private _getHost(): Node | undefined {
    const { hostId } = this.props;
    const doc = getDocument(this._rootElement.current);

    if (!doc) {
      return undefined;
    }

    if (hostId) {
      return doc.getElementById(hostId) as Node;
    } else {
      return _defaultHostSelector ? doc.querySelector(_defaultHostSelector) as Node : doc.body;
    }
  }

}
