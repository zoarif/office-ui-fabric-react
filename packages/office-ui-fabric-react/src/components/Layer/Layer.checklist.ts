import { ChecklistStatus } from '../../demo/ComponentStatus/ComponentStatus.types';

export const LayerStatus = {
  keyboardAccessibilitySupport: ChecklistStatus.unknown,
  markupSupport: ChecklistStatus.unknown,
  highContrastSupport: ChecklistStatus.pass,
  rtlSupport: ChecklistStatus.pass,
  testCoverage: ChecklistStatus.good
};