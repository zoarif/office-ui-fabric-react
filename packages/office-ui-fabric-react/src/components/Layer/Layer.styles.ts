import { ILayerStyleProps, ILayerStyles } from './Layer.types';
import { getGlobalClassNames } from '../../Styling';

const GlobalClassNames = {
  root: 'ms-Layer',
  rootNoHost: 'ms-Layer--fixed',
  content: 'ms-Layer-content',
};

export const getStyles = (
  props: ILayerStyleProps
): ILayerStyles => {
  const {
    className,
    isNotHost,
    theme
  } = props;

  const classNames = getGlobalClassNames(GlobalClassNames, theme);

  return ({
    root: [
      classNames.root,
      isNotHost && [
        classNames.rootNoHost,
        {
          position: 'fixed',
          zIndex: 1000000,
          top: 0,
          left: 0,
          width: '100vw',
          height: '100vh',
          visibility: 'hidden'
        }
      ],
      className
    ],
    content: [
      classNames.content,
      {
        visibility: 'visible'
      }
    ]
  });
};
