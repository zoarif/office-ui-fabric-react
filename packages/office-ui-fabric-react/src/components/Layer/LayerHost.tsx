import * as React from 'react';
import {
  BaseComponent,
  css
} from '../../Utilities';
import { LayerBase } from './Layer.base';
import { ILayerHostProps } from './LayerHost.types';

export class LayerHost extends BaseComponent<ILayerHostProps, {}> {

  public shouldComponentUpdate() {
    return false;
  }

  public componentDidMount(): void {
    LayerBase.notifyHostChanged(this.props.id!);
  }

  public componentWillUnmount(): void {
    LayerBase.notifyHostChanged(this.props.id!);
  }

  public render(): JSX.Element {
    return (
      <div { ...this.props } className={ css('ms-LayerHost', this.props.className) } />
    );
  }
}