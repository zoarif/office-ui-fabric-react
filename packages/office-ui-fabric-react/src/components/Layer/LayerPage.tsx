
import * as React from 'react';
import {
  ExampleCard,
  IComponentDemoPageProps,
  ComponentPage,
  PageMarkdown,
  PropertiesTableSet
} from '@uifabric/example-app-base';
import { LayerBasicExample } from './examples/Layer.Basic.Example';
import { LayerHostedExample } from './examples/Layer.Hosted.Example';
import { LayerCustomizedExample } from './examples/Layer.Customized.Example';
import { ComponentStatus } from '../../demo/ComponentStatus/ComponentStatus';
import { LayerStatus } from './Layer.checklist';

const LayerBasicExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Layer/examples/Layer.Basic.Example.tsx') as string;
const LayerHostedExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Layer/examples/Layer.Hosted.Example.tsx') as string;
const LayerCustomizedExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Layer/examples/Layer.Customized.Example.tsx') as string;

export class LayerPage extends React.Component<IComponentDemoPageProps, {}> {
  public render(): JSX.Element {
    return (
      <ComponentPage
        title='Layer'
        componentName='LayerExample'
        componentUrl='https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/Layer'
        exampleCards={
          <div>
            <ExampleCard title='Basic layered content' code={ LayerBasicExampleCode }>
              <LayerBasicExample />
            </ExampleCard>
            <ExampleCard title='Using LayerHost to control projection' code={ LayerHostedExampleCode }>
              <LayerHostedExample />
            </ExampleCard>
            <ExampleCard title='Using Customizer to control the default layer behavior' code={ LayerCustomizedExampleCode }>
              <LayerCustomizedExample />
            </ExampleCard>
          </div>
        }
        propertiesTables={
          <PropertiesTableSet
            sources={ [
              require<string>('!raw-loader!office-ui-fabric-react/src/components/Layer/Layer.types.ts')
            ] }
          />
        }
        overview={
          <PageMarkdown>
            { require<string>('!raw-loader!office-ui-fabric-react/src/components/Layer/docs/LayerOverview.md') }
          </PageMarkdown>
        }
        bestPractices={
          <div />
        }
        dos={
          <PageMarkdown>
            { require<string>('!raw-loader!office-ui-fabric-react/src/components/Layer/docs/LayerDos.md') }
          </PageMarkdown>
        }
        donts={
          <PageMarkdown>
            { require<string>('!raw-loader!office-ui-fabric-react/src/components/Layer/docs/LayerDonts.md') }
          </PageMarkdown>
        }
        isHeaderVisible={ this.props.isHeaderVisible }
        componentStatus={
          <ComponentStatus
            { ...LayerStatus }
          />
        }
      />
    );
  }
}
