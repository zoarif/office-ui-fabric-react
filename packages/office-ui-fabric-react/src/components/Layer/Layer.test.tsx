/* tslint:disable:no-unused-variable */
import * as React from 'react';
import * as ReactDOM from 'react-dom';
import * as PropTypes from 'prop-types';
/* tslint:enable:no-unused-variable */
import * as renderer from 'react-test-renderer';

import { Layer } from './Layer';
import { LayerHost } from './LayerHost';

describe('Layer', () => {
  it('renders Layer correctly', () => {
    const component = renderer.create(<Layer>Content</Layer>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('can render in a targeted LayerHost and pass context through', () => {

    class Child extends React.Component<{}, {}> {
      public static contextTypes = {
        foo: PropTypes.string.isRequired
      };

      public context: any;

      public render(): JSX.Element {
        return (
          <div id='child'>{ this.context.foo }</div>
        );
      }
    }

    class Parent extends React.Component<{}, {}> {
      public static childContextTypes = {
        foo: PropTypes.string
      };

      public getChildContext() {
        return {
          foo: 'foo'
        };
      }

      public render(): JSX.Element {
        return (
          <div id='parent'>
            <Layer hostId='foo'>
              <Child />
            </Layer>
          </div>
        );
      }
    }

    class App extends React.Component<{}, {}> {

      public render(): JSX.Element {
        return (
          <div id='app'>
            <Parent />
            <LayerHost id='foo' />
          </div>
        );
      }
    }

    const appElement = document.createElement('div');

    try {
      document.body.appendChild(appElement);
      ReactDOM.render(<App />, appElement);

      const parentElement = appElement.querySelector('#parent');

      expect(parentElement).toBeDefined();
      expect(parentElement!.ownerDocument).toBeDefined();

      const childElement = appElement.querySelector('#child') as Element;

      expect(childElement.textContent).toEqual('foo');
    } finally {
      ReactDOM.unmountComponentAtNode(appElement);
      appElement.remove();
    }
  });

});