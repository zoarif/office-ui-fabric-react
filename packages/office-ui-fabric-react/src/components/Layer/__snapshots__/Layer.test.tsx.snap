// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layer renders Layer correctly 1`] = `
<span
  className="ms-Layer"
/>
`;
